### PR TITLE
Implement native support StringView for find in set

### LINKED ISSUE #11954: Update `FIND_IN_SET` scalar function to support Utf8View
Part of https://github.com/apache/datafusion/issues/11752 and https://github.com/apache/datafusion/issues/11790

Currently, a call to `FIND_IN_SET` with a Utf8View datatypes induces a cast. After the change that fixes this issue, it should not.

```sql
query TT
EXPLAIN SELECT
  FIND_IN_SET(column1_utf8view, 'a,b,c,d') as c
FROM test;
----
logical_plan
01)Projection: find_in_set(CAST(test.column1_utf8view AS Utf8), Utf8("a,b,c,d")) AS c
02)--TableScan: test projection=[column1_utf8view]
```

`find_in_set` is defined here: https://github.com/apache/datafusion/blob/main/datafusion/functions/src/unicode/find_in_set.rs

casting tests are in: https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/string_view.slt

### Is your feature request related to a problem or challenge?

We are working to add complete StringView support in DataFusion, which permits potentially much faster processing of string data. See https://github.com/apache/datafusion/issues/10918 for more background.

Today, most DataFusion string functions support DataType::Utf8 and DataType::LargeUtf8 and when called with a StringView argument DataFusion will cast the argument back to DataType::Utf8 which is expensive.

To realize the full speed of StringView, we need to ensure that all string functions support the DataType::Utf8View directly.

### Describe the solution you'd like

Update the function to support DataType::Utf8View directly

### Describe alternatives you've considered

The typical steps are:
1. Write some tests showing the function doesn't support Utf8View (see the tests in [`string_view.slt`](https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/string_view.slt) to ensure the arguments are not being cast
2. Change the `Signature` of the function to accept `Utf8View` in addition to `Utf8`/`LargeUtf8`
3. Update the implementation of the function to operate on `Utf8View`

Example PRs
* Update to use an arrow kernel that already supports StringView:  https://github.com/apache/datafusion/pull/11787
* Change the implementation to support StringView directly: https://github.com/apache/datafusion/pull/11676
* Change implementation (option 2): https://github.com/apache/datafusion/pull/11556

### Additional context

The documentation of string functions can be found here: https://datafusion.apache.org/user-guide/sql/scalar_functions.html#string-functions

To test a function with StringView with `datafusion-cli` you can use an example like this (replacing `starts_with` with the relevant function)
```sql
> create table foo as values (arrow_cast('foo', 'Utf8View'), arrow_cast('bar', 'Utf8View'));
0 row(s) fetched.
Elapsed 0.043 seconds.

> select starts_with(column1, column2) from foo;
+--------------------------------------+
| starts_with(foo.column1,foo.column2) |
+--------------------------------------+
| false                                |
+--------------------------------------+
1 row(s) fetched.
Elapsed 0.015 seconds.
```

To see if it is using utf8 view, use `EXPLAIN` to see the plan and verify there is no `CAST`. In this example the `CAST(column1@0 AS Utf8)` indicates that the function is not using `Utf8View` natively

```sql
> explain select starts_with(column1, column2) from foo;
+---------------+------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                         |
+---------------+------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: starts_with(CAST(foo.column1 AS Utf8), CAST(foo.column2 AS Utf8))                                                |
|               |   TableScan: foo projection=[column1, column2]                                                                               |
| physical_plan | ProjectionExec: expr=[starts_with(CAST(column1@0 AS Utf8), CAST(column2@1 AS Utf8)) as starts_with(foo.column1,foo.column2)] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                                                              |
|               |                                                                                                                              |
+---------------+------------------------------------------------------------------------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.006 seconds.
```

It is also often good to test with a constant as well (likewise there should be no cast):

```sql
> explain select starts_with(column1, 'foo') from foo;
+---------------+----------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                     |
+---------------+----------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: starts_with(CAST(foo.column1 AS Utf8), Utf8("foo"))                                          |
|               |   TableScan: foo projection=[column1]                                                                    |
| physical_plan | ProjectionExec: expr=[starts_with(CAST(column1@0 AS Utf8), foo) as starts_with(foo.column1,Utf8("foo"))] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                                          |
|               |                                                                                                          |
+---------------+----------------------------------------------------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.002 seconds.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by PsiACE**
## Which issue does this PR close?

Closes #11954 .

## Rationale for this change

Ref: https://github.com/apache/datafusion/blob/main/datafusion/functions/src/unicode/character_length.rs

### CODE REVIEW

File: datafusion/sqllogictest/test_files/string_view.slt [hunk 1]
```diff
@@ -913,7 +913,7 @@ EXPLAIN SELECT
 FROM test;
 ----
 logical_plan
-01)Projection: find_in_set(CAST(test.column1_utf8view AS Utf8), Utf8("a,b,c,d")) AS c
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/string_view.slt:916]: I think we should probably add a test that actually invokes `find_in_set` with a `StringVieAwArray` argument.

However, I think @2010YOUY01  has some ideas on a more general testing framework, so maybe this is good enough for now. Let me know what you think

https://github.com/apache/datafusion/issues/11790#issuecomment-2284100621

**Review by alamb — APPROVED** [review]
Thanks @PsiACE  -- I think we should consider a short test that actually invokes the function but otherwise looks great to me ❤️

File: datafusion/sqllogictest/test_files/string_view.slt (hunk 1 above)
Comment (PsiACE) [reply datafusion/sqllogictest/test_files/string_view.slt:916]: I will add some sqllogictests to make sure it is covered.

Comment (PsiACE) [reply datafusion/sqllogictest/test_files/string_view.slt:916]: Done! https://github.com/apache/datafusion/pull/11970/commits/5d5a4e6ab09bcef4619eff2557c3ee71d37da2bd

**Review by alamb — APPROVED** [review]
Thanks again @PsiACE

**Pull request merged**